src/Microsoft.Identity.Web/Resource/TokenAuthorizationRules.cs

```diff
@@ -0,0 +1,150 @@
+
+            // Does the web API demand specific tenants?
+            if (tokenAuthorizationOptions.AllowedTenantIds != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Tid, ClaimConstants.TenantId },
+                    tokenAuthorizationOptions.AllowedTenantIds,
+                    IDWebErrorMessage.ApiAllowedTenantsCantBeComputed,
+                    IDWebErrorMessage.ApiTenantNotAllowed);
+            }
+
+            // Does the web API demand specific clients (oid)
+            if (tokenAuthorizationOptions.AllowedClientApplications != null)
+            {
+                VerifyClaimIsInAcceptedValues(
+                    claimsPrincipal,
+                    new[] { ClaimConstants.Oid, ClaimConstants.ObjectId },
+                    tokenAuthorizationOptions.AllowedClientApplications,
+                    IDWebErrorMessage.ApiAllowedClientApplicationsCantBeComputed,
+                    IDWebErrorMessage.ApiClientApplicationNotAllowed);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimEquals(
+            ClaimsPrincipal claimsPrincipal,
+            string claimName,
+            string expectedValue,
+            string messageWhenClaimsNotFound,
+            string messageWhenUnexpectedClaimValue)
+        {
+            string? claimValue = ClaimsPrincipalExtensions.GetClaimValue(claimsPrincipal, claimName);
+            if (claimValue == null)
+            {
+                throw new UnauthorizedAccessException(messageWhenClaimsNotFound);
+            }
+
+            if (string.Equals(claimValue, expectedValue, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new UnauthorizedAccessException(messageWhenUnexpectedClaimValue);
+            }
+        }
+
+        /// <summary>
+        /// Verify the value of a claim.
+        /// </summary>
+        /// <param name="claimsPrincipal">Claims.</param>
+        /// <param name="claimName">Name (type) of the claim to verify.</param>
+        /// <param name="expectedValue">Expected value.</param>
+        /// <param name="messageWhenClaimsNotFound">Error message thrown when the claim is not found.</param>
+        /// <param name="messageWhenUnexpectedClaimValue">Error message thrown when claim value is not the expected one.</param>
+        private static void VerifyClaimDoesNotExistOrDoesNotEquals(
```

Comment:
Actually, I'm not sure. If we want to have verification that we don't accept app-only tokens, the only way to do it is:
- either the claim does not exist (as it's not generated if the token is not an app-only token)
- or it exists and its value is "app"
Or am I missing something, @psignoret @hpsin ?

Reply:
If we wanted to demand an app-only token, then we'd need to have the claim and it's value should be app.

Reply:
This method is used when we _don't_ want to allow app-only tokens (because `ApiAllowsAppOnlyTokens` is `false`). To know for sure, we need to verify that the `idtyp` claim is present, AND we need it's value to _not_ be "app". So, we should throw an exception if the claim is not present, OR if it's present and the value is "app".